File: 100 The 3n + 1 Problem/main.py
def cycle(memo, n):
    if n in memo:
        return memo[n]
    if n == 1:
        return 1
    memo[n] = 1 + cycle(memo, 3*n + 1 if n & 1 == 1 else n >> 1)
    return memo[n]

def solve(memo, i, j):
    res = 0
    for n in range(i, j + 1):
        res = max(res, cycle(memo, n))
    return res

if __name__ == '__main__':
    memo = {}
    while True:
        try:
            i, j = map(int, input().split())
        except EOFError:
            break
            
        ans = solve(memo, min(i, j), max(i, j))
        print('{} {} {}'.format(i, j, ans))

File: 100 The 3n + 1 Problem/main.cpp
#include <iostream>
#include <unordered_map>
#include <queue>

using namespace std;

int Cycle(const int num, unordered_map<int, int>& m) {
    if (m.find(num) == m.end()) {
        queue<int> q;
        auto tmp = num;
        while (true) {
            q.push(tmp);
            if (tmp == 1) break;
            if (tmp & 1) {
                tmp = 3 * tmp + 1;
            } else {
                tmp >>= 1;
            }
        }
        while(!q.empty()) {
            m[q.front()] = q.size();
            q.pop();
        }
    }
    return m[num];
}

int Solve(int i, int j, unordered_map<int, int>& m) {
    int res = 0;
    for (int k = i; k <= j; ++k) {
        res = max(res, Cycle(k, m));
    }
    return res;
}

int main() {
    unordered_map<int, int> m;
    int i, j;
    while(cin >> i >> j) {
        int i2, j2;
        i2 = i <= j ? i : j;
        j2 = i <= j ? j : i;
        auto ans = Solve(i2, j2, m);
        cout << i << " " << j << " " << ans << "\n";
    }
    return EXIT_SUCCESS;
}
